packages/firestore/src/api/database.ts

```diff
@@ -1000,43 +1000,91 @@ export class DocumentReference implements firestore.DocumentReference {
     };
   }
 
-  get(): Promise<firestore.DocumentSnapshot> {
-    validateExactNumberOfArgs('DocumentReference.get', arguments, 0);
+  get(options?: firestore.GetOptions): Promise<firestore.DocumentSnapshot> {
+    validateOptionNames('DocumentReference.get', options, ['source']);
+    if (options) {
+      validateNamedOptionalPropertyEquals(
+        'DocumentReference.get',
+        'options',
+        'source',
+        options.source,
+        ['default', 'server', 'cache']
+      );
+    }
     return new Promise(
       (resolve: Resolver<firestore.DocumentSnapshot>, reject: Rejecter) => {
-        const unlisten = this.onSnapshotInternal(
-          {
-            includeQueryMetadataChanges: true,
-            includeDocumentMetadataChanges: true,
-            waitForSyncWhenOnline: true
-          },
-          {
-            next: (snap: firestore.DocumentSnapshot) => {
-              // Remove query first before passing event to user to avoid
-              // user actions affecting the now stale query.
-              unlisten();
-
-              if (!snap.exists && snap.metadata.fromCache) {
-                // TODO(dimond): If we're online and the document doesn't
-                // exist then we resolve with a doc.exists set to false. If
-                // we're offline however, we reject the Promise in this
-                // case. Two options: 1) Cache the negative response from
-                // the server so we can deliver that even when you're
-                // offline 2) Actually reject the Promise in the online case
-                // if the document doesn't exist.
-                reject(
-                  new FirestoreError(
-                    Code.ABORTED,
-                    'Failed to get document because the client is ' + 'offline.'
-                  )
-                );
-              } else {
-                resolve(snap);
-              }
-            },
-            error: reject
+        if (options && options.source === 'cache') {
+          this.firestore
+            .ensureClientConfigured()
+            .getDocumentFromLocalCache(this._key)
+            .then((doc: Document) => {
+              resolve(
+                new DocumentSnapshot(
+                  this.firestore,
+                  this._key,
+                  doc,
+                  /*fromCache=*/ true
+                )
+              );
+            }, reject);
+        } else {
+          this.getViaSnapshotListener_(resolve, reject, options);
+        }
+      }
+    );
+  }
+
+  private getViaSnapshotListener_(
```

Comment:
I don't think we usually add `_` to private methods.

We would ideally not use `_` as a naming convention at all and just rely on private/public modifiers, etc but with our current build setup, any properties or public methods on classes end up showing up on the object at runtime and so end users could potentially discover them.  So for any properties / public methods we don't want end users to see / use (because they're not part of our official API), we add a `_` suffix which triggers the minification step in our build to obfuscate the name: https://github.com/firebase/firebase-js-sdk/blob/master/packages/firebase/webpack.config.js#L48

Anyway, private methods are completely hidden at runtime based on how TypeScript implements private methods and so we don't need / want to use the `_` suffix (here and on the other getViaSnapshotListener_ method).